fiftyone/core/utils.py

```diff
@@ -1144,10 +1144,317 @@ def _compute_batch_size(self):
+            time_delta = current_time - self._last_time
+
+        self._last_time = current_time
+        return time_delta
+
+
+# Define this for backwards compatibility in case someone was using this
+# batcher directly
+DynamicBatcher = LatencyDynamicBatcher
+
+
+class BsonSizeDynamicBatcher(BaseDynamicBatcher):
+    """Class for iterating over the elements of an iterable with a dynamic
+    batch size to achieve a desired BSON content size.
+
+    The batch sizes emitted when iterating over this object are dynamically
+    scaled such that the total BSON size of the batch is as close as
+    possible to a specified target size.
+
+    This batcher requires that backpressure feedback be manually provided,
+    so that the provided list is BSON-able in order to calculate an accurate
+    estimate of total batch content size to adjust next batch size.
+
+    This class is often used in conjunction with a :class:`ProgressBar` to keep
+    the user appraised on the status of a long-running task.
+
+    Example usage::
+
+        import fiftyone.core.utils as fou
+
+        elements = range(int(1e7))
+
+        batcher = fou.BsonSizeDynamicBatcher(
+            elements, target_size=2**20, max_batch_beta=2.0
+        )
+
+        # Raises ValueError after first batch, we forgot to apply backpressure
+        for batch in batcher:
+            print("batch size: %d" % len(batch))
+
+        # Now it works
+        for batch in batcher:
+            print("batch size: %d" % len(batch))
+            batcher.apply_backpressure(batch)
+
+        batcher = fou.BsonSizeDynamicBatcher(
+            elements,
+            target_size=2**20,
+            max_batch_beta=2.0,
+            progress=True
+        )
+
+        with batcher:
+            for batch in batcher:
+                print("batch size: %d" % len(batch))
+                batcher.apply_backpressure(batch)
+
+    Args:
+        iterable: an iterable
+        target_size (1048576): the target batch bson content size, in bytes
+        init_batch_size (1): the initial batch size to use
+        min_batch_size (1): the minimum allowed batch size
+        max_batch_size (None): an optional maximum allowed batch size
+        max_batch_beta (None): an optional lower/upper bound on the ratio
+            between successive batch sizes
+        return_views (False): whether to return each batch as a
+            :class:`fiftyone.core.view.DatasetView`. Only applicable when the
+            iterable is a :class:`fiftyone.core.collections.SampleCollection`
+        progress (False): whether to render a progress bar tracking the
+            consumption of the batches
+        total (None): the length of ``iterable``. Only applicable when
+            ``progress=True``. If not provided, it is computed via
+            ``len(iterable)``, if possible
+    """
+
+    manual_backpressure = True
+
+    def __init__(
+        self,
+        iterable,
+        target_size=2**15,
```

Comment:
nit: I see 2**15 here but in `fo.config` the default is `2**20`. Standardize on `2**20`?

Reply:
relic of testing different values